propagation/http_trace_context_propagator_test.go

```diff
@@ -0,0 +1,152 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package propagation_test
+
+import (
+	"context"
+	"net/http"
+	"testing"
+
+	"go.opentelemetry.io/api/trace"
+
+	"github.com/google/go-cmp/cmp"
+
+	"go.opentelemetry.io/api/core"
+	mocktrace "go.opentelemetry.io/internal/trace"
+	"go.opentelemetry.io/propagation"
+)
+
+var (
+	traceID = core.TraceID{High: 0x4bf92f3577b34da6, Low: 0xa3ce929d0e0e4736}
+	spanID  = uint64(0x00f067aa0ba902b7)
+)
+
+func TestExtractTraceContextFromHTTPReq(t *testing.T) {
+	trace.SetGlobalTracer(&mocktrace.MockTracer{})
+	propagator := propagation.HttpTraceContextPropagator()
+	tests := []struct {
+		name   string
+		header string
+		wantSc core.SpanContext
+	}{
+		{
+			name:   "future version",
+			header: "02-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01",
+			wantSc: core.SpanContext{
+				TraceID:      traceID,
+				SpanID:       spanID,
+				TraceOptions: core.TraceOptionSampled,
+			},
+		},
+		{
+			name:   "zero trace ID and span ID",
+			header: "00-00000000000000000000000000000000-0000000000000000-01",
+			wantSc: core.EmptySpanContext(),
+		},
+		{
+			name:   "valid header",
+			header: "00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01",
+			wantSc: core.SpanContext{
+				TraceID:      traceID,
+				SpanID:       spanID,
+				TraceOptions: core.TraceOptionSampled,
+			},
+		},
+		{
+			name:   "missing options",
+			header: "00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7",
+			wantSc: core.EmptySpanContext(),
+		},
+		{
+			name:   "empty options",
+			header: "00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-",
+			wantSc: core.EmptySpanContext(),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			req, _ := http.NewRequest("GET", "http://example.com", nil)
+			req.Header.Set("traceparent", tt.header)
+
+			ctx := context.Background()
+			gotSc := propagator.Extract(ctx, req.Header)
+			if diff := cmp.Diff(gotSc, tt.wantSc); diff != "" {
+				t.Errorf("Extract Tracecontext: %s: -got +want %s", tt.name, diff)
+			}
+		})
+	}
+}
+
+func TestInjectTraceContextToHTTPReq(t *testing.T) {
+	var id uint64
+	mockTracer := &mocktrace.MockTracer{
+		Sampled:     false,
+		StartSpanId: &id,
+	}
+	propagator := propagation.HttpTraceContextPropagator()
+	tests := []struct {
+		name       string
+		sc         core.SpanContext
+		wantHeader string
+	}{
+		{
+			name: "valid spancontext, sampled",
+			sc: core.SpanContext{
+				TraceID:      traceID,
+				SpanID:       spanID,
+				TraceOptions: core.TraceOptionSampled,
+			},
+			wantHeader: "00-4bf92f3577b34da6a3ce929d0e0e4736-0000000000000001-01",
+		},
+		{
+			name: "valid spancontext, not sampled",
+			sc: core.SpanContext{
+				TraceID: traceID,
+				SpanID:  spanID,
+			},
+			wantHeader: "00-4bf92f3577b34da6a3ce929d0e0e4736-0000000000000002-00",
+		},
+		{
```

Comment:
I wonder if there should be a test case with a valid span context having options like `0xff`, so the expected header should have `01` as options, not `ff`.

Reply:
with version '0' I added test to check that option should not be >2. For future version I added other values of option but expected value to be 01 or 00.

Reply:
I can't see those tests anywhere.

Reply:
for version 0, anything other than 01 should fail
```
		{
			name:   "trace-flag unused bits set",
			header: "00-ab000000000000000000000000000000-cd00000000000000-09",
		},
```
For future version accept 01 and 00
```
		{
			name:   "future options with sampled bit set",
			header: "02-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-09",
			wantSc: core.SpanContext{
				TraceID:      traceID,
				SpanID:       spanID,
				TraceOptions: core.TraceOptionSampled,
			},
		},
		{
			name:   "future options with sampled bit cleared",
			header: "02-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-08",
			wantSc: core.SpanContext{
				TraceID: traceID,
				SpanID:  spanID,
			},
		},
```

Reply:
Oh, I meant the test for the inject part, where as input you have a span context like:

```go
core.SpanContext{
	TraceID:      core.TraceID{High: 0x4bf92f3577b34da6, Low: 0xa3ce929d0e0e4736},
	SpanID:       uint64(0x00f067aa0ba902b7),
	TraceOptions: 0xff,
}
```

and the expected generated "traceparent" HTTP header (for now) would be `00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01` (not `ff` at the end).

Reply:
added.